lib/src/bootloader.rs

```diff
@@ -26,3 +26,113 @@ pub(crate) fn install_via_bootupd(
+        filesystems: Vec<Filesystem>,
+    }
+
+    let findmnt_desc = format!("Running findmnt to identify the boot partition with UUID={boot_uuid}");
+    let findmnt_task = Task::new(findmnt_desc, "findmnt");
+    let findmnt_args = ["--json", "--output", "TARGET,SOURCE,FSTYPE",
+                        "--uniq", "--first-only", "--nofsroot",
+                        "--source", &(format!("UUID={boot_uuid}"))];
+    let findmnt_out  = findmnt_task.args(findmnt_args).quiet().read()?;
+    let findmnt_json = serde_json::from_str::<Findmnt>(&findmnt_out).context("parsing findmnt output")?;
+
+    let boot_fs = findmnt_json.filesystems.get(0).context("finding boot filesystem")?;
+    let boot_dev = Utf8Path::new(&boot_fs.source);
+    let boot_dir = Utf8Path::new(&boot_fs.target);
+
+    // Run sfdisk to determine the boot partition offset
+
+    #[derive(serde::Deserialize)]
+    struct Partition {
+        node:  String,
+        start: i64,
+    }
+
+    #[derive(serde::Deserialize)]
+    struct PartitionTable {
+        sectorsize: i64,
+        partitions: Vec<Partition>,
+    }
+
+    #[derive(serde::Deserialize)]
+    struct Sfdisk {
+        partitiontable: PartitionTable
+    }
+
+    let sfdisk_desc = format!("running sfdisk to determine the partition offset of {boot_dev}");
+    let sfdisk_task = Task::new(sfdisk_desc, "sfdisk");
+    let sfdisk_args = ["--json", device.as_str()];
+    let sfdisk_out  = sfdisk_task.args(sfdisk_args).quiet().read()?;
+    let sfdisk_json = serde_json::from_str::<Sfdisk>(&sfdisk_out).context("parsing findmnt output")?;
+
+    let block_size  = sfdisk_json.partitiontable.sectorsize;
+    let partitions  = sfdisk_json.partitiontable.partitions;
+    let boot_part   = partitions.iter().find(|p| p.node == boot_dev).context("finding boot partition")?;
+    let boot_offset = boot_part.start;
+
+    // Find the default BLS conf under /boot/loader/entries
+
+    let bls_dir = boot_dir.join("boot/loader/entries");
+
+    let ls_desc = format!("running ls to find the default BLS configuration");
+    let ls_task = Task::new(ls_desc, "ls");
+    let ls_args = ["--sort=version", "--reverse", bls_dir.as_str()];
+    let ls_out = ls_task.args(ls_args).quiet().read()?;
+
+    let bls_name = ls_out.lines().find(|filename| filename.ends_with(".conf")).context("finding BLS conf")?;
+    let bls_path = bls_dir.join(bls_name);
+    let bls_conf = std::fs::read_to_string(&bls_path).context(format!("reading {bls_path:?}"))?;
+
+    let mut image = "";
+    let mut ramdisk = "";
+    let mut parameters = "";
+
+    for line in bls_conf.lines() {
+        match line.split_once(char::is_whitespace) {
+            Some(("linux",   val)) => image      = val.trim(),
+            Some(("initrd",  val)) => ramdisk    = val.trim(),
+            Some(("options", val)) => parameters = val.trim(),
+            _ => (),
+        }
+    };
+
+    let image_path   = boot_dir.join(image.trim_start_matches('/')).canonicalize_utf8()?;
```

Comment:
Did something specifically motivate the `.canonicalize_utf8()` here out of curiosity?

Reply:
`canonicalize_utf8` is just a cosmetic, nice-to-have thing here.

When the target filesystem is `/boot`, `image_path` becomes something like `/run/bootc/mounts/rootfs/boot/boot/ostree/...`. `boot` directory appears twice in the path, and the second `boot` directory is actually a symlink to `.`.

This path representation is OK for zipl to work correctly, but looks a little bit strange. So I use `canonicalize_utf8` to remove the duplicated `boot` from the path.